### ISSUE TITLE
Image2Image GAN training curves are not going as expected

### ISSUE THREAD

**Issue by tmquan** [ppwwyyxx/tensorpack#241]
Hi tensorpack, 

Did you have a chance to look at the accuracies and the losses when training pix2pix example?
I have tried it with many datasets provided in the papers but it seems that the discriminator curves of accuracy are not going down to 0.5 and the generator curves of accuracy are not going up to 1.0 to match the theory of GAN. 

![image](https://cloud.githubusercontent.com/assets/4506781/25462315/291b7586-2b29-11e7-93df-f67c9c44b9a0.png)

Bests

**Comment by ppwwyyxx** [comment]
The "theory" doesn't necessarily work in practice.

**Comment by tmquan** [comment]
I just reopened for more discussion. 
If this is the case, we can not trust the result from the generator, can we?
The generated images are just qualitatively fine, not quantitatively.

**Comment by tmquan** [comment]
In the GAN.py implementation, is it true that we need to minimize d_loss, as opposed to the **minimizing -d_loss** (maximizing d_loss) ops in GAN theory?

```python
# by default, run one d_min after one g_min
        self.g_min = opt.minimize(self.model.g_loss, var_list=self.model.g_vars, name='g_op')
        with tf.control_dependencies([self.g_min]):
            self.d_min = opt.minimize(self.model.d_loss, var_list=self.model.d_vars, name='d_op')
        self.train_op = self.d_min
```

**Comment by ppwwyyxx** [comment]
In theory it maximizes the likelyhood not the loss.

**Comment by tmquan** [comment]
Thank @ppwwyyxx ,

I have another question. Why did you need tf.control_dependencies before d_min ops?

**Comment by PatWie** [comment]
You can also write
```
sess.run(self.d_min)
sess.run(self.g_min)
```

later in the GANTrainer. But the 
http://devdocs.io/tensorflow~python/tf/control_dependencies
connect both operations within the graph, which reduces the sess.run overhead.

**Comment by Skylion007** [comment]
@PatWie Could you exploit this fact to make a Sync / Async MultiGPU GAN trainer? I've been trying to figure out how to add one in, but I have had little luck figuring out the best way to go about it.

**Comment by ppwwyyxx** [comment]
Another difference is that, with
```
sess.run(self.d_min)
sess.run(self.g_min)
```
G and D are trained on different batch of input images. When written together they are trained on same batch of images.

**Comment by yhlleo** [comment]
I found an interesting thing that if I change the input data:

`input, output = input / 128.0 - 1, output / 128.0 - 1`

as:

`output = output / 128.0 - 1`.

Then, both `discrim/accuracy` and `gen/accuracy` tend to be `0.5` during the training.

Therefore, the operation `input = input / 128.0 - 1` is unnecessary.

**Comment by PatWie** [comment]
But this only works because of BatchNorm and **leaky**ReLU.

**Comment by ppwwyyxx** [comment]
@Skylion007 just made a MultiGPU GAN trainer in 01486c39dfbe141adbff3. 
Tested on BEGAN, it scales 1.95x on 2 GPUs and 3.88x on 4 GPUs.
May not be the most efficient way but the performance is OK.
